[알림] 알림 연관 기능 구현(#66)

Minjae-An opened the pull request:
## 작업 대상
다음 기능들 구현
- 알림 동의 내역 조회 
- 알림 on/off 설정(알림 동의 내역 업데이트 및 생성) 
- 최근 받은 알림 목록 조회
- 읽지 않은 알림 존재 여부 조회
- device token 업데이트(생성)
- FCM 이용, 푸시 알림 전송

## 📄 작업 내용
<!-- 작업한 내용을 간단히 요약해주세요. -->
- 알림 연관 로직 구현

## 🙋🏻 주의 사항
<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->
### 최근 받은 알림 목록 조회 API가 `@PatchMapping` 인 이유
- 해당 API는 생성 일시 내림차순 기준으로 최근 받은 알림 9개를 페이지 형태 응답으로 제공
- 알림 목록을 조회할 때(이 API에 요청할 때) 알림을 읽음 처리하는 것으로 요구사항 정의
- 따라서 이 API에 요청할 경우, 알림 목록 조회와 알림 읽음 처리 두 작업이 수행됨
- 리소스의 일부분을 수정하는 작업을 수행한다는 측면에서 PATCH로 설정

### 알림 읽음 처리 로직 설명
- 알림 목록 조회시 응답 되는 알림 중 가장 최근에 받은(가장 큰) 알림 ID 도출
- 해당 ID보다 이전(작거나 같은 ID를 가진) 알림들 전부 읽음 처리
- QueryDSL의 벌크성 수정 쿼리를 통해 해당 작업 수행
- 벌크성 수정 쿼리는 영속성 컨텍스트를 무시하고 수행됨
- DB와 컨텍스트 데이터 동기화를 위해 해당 쿼리 수행 후 컨텍스트 초기화 과정 수행

### badge = 최근 읽지 않은 알림 개수
- ios 홈화면에서 특정 앱 아이콘 위에 사용자가 확인하지 않은 알림, 메시지 등을 표시하는 기능을 badge라 지칭
- 우리 서비스의 경우 badge = 최근 읽지 않은 알림 개수
- ios에서 badge 기능을 구현하기 위해서는 서버에서 사용자별 badge를 관리, 푸시 알림시 포함하여 전달해야 함
- 이에 따른 비즈니스 로직 구현

### 알림 핸들링을 SpringEvent를 활용하여 구현한 이유?
- 알림 자체는 알림을 생성하게 되는 비즈니스 로직과 밀접하지 않은 편의 기능
- 따라서 해당 비즈니스 로직 수행에 알림 핸들링 로직이 영향을 주지 않는, 결합도가 낮은 형태로 구현되어야 한다 생각
- 이에 따라 스프링 이벤트를 활용하여 로직을 분리, 결합도를 낮춤

### 알림 이벤트 로직의 구성
- 알림 종류에 따라 알림 제목과 내용이 다를 뿐 전달되어야 하는 데이터는 다음으로 같음
    - 알림 종류(ex. 화살 보내기, 채팅 요청)
    - 알림 생성 사용자 ID
    - 알림 받는 사용자 badge(읽지 않은 알림 개수)
    - 알림 생성 일시
- 이에 따라 알림 종류를 enum으로 정의하고 종류에 따라 다른 제목, 내용을 필드로 관리
- 알림 이벤트에는 알림 내역 생성, 푸시 알림 전송에 필요한 알림 타입, 알림 전송자, 알림 수신자, 
  알림 생성 일시를 포함하도록 구성
- 알림 이벤트는 발생하는 순간, 불변 상태로 유지되어야 하므로 Java 17에서 제공하는 `record` 활용
- 이 이벤트 객체를 활용해 푸시 알림 전송과 알림 내역 생성 및 저장을 처리하도록 로직 구현
- 상황에 따라 이벤트 객체를 별도로 정의하거나, 알림 내역 생성 및 저장을 위한 요청 DTO를 별도로 정의하는 것은
  현재 상황에선 불필요하게 복잡도를 높인다고 판단

### FCM 서버에 푸시 알림을 요청하는 로직을 비동기 처리한 이유?
- FCM은 외부 서비스로 해당 서비스 성능이 코어 서비스 성능에 영향을 주는 것은 부적합하다고 생각
- 따라서 FCM 서버에 푸시 알림을 요청하는 로직을 비동기 처리
- 같은 맥락에서 해당 로직에서 발생하는 예외가 코어 비즈니스 로직에 전파되어선 안 된다고 생각
- 이에 따라 발생하는 예외를 로그로 남겨 대응할 수 있도록 구성

## 📎 관련 이슈
### JSON 파일을 secrets으로 관리하는 방법
- github secrets 공식 문서에서는 json 파일을 직접 secrets으로 저장하여 사용하는 형태를 권장하지 않음
- 실제로 fcm-account-key json 파일을 secrets에 그대로 복사하여 저장할 경우 actions 빌드 수행 과정서 
  제대로 인식되지 않음
- 로컬에서 해당 파일을 base64 인코딩하여 secrets에 저장하고 actions 수행 과정에서 디코딩하여 사용하는 
  형식으로 문제 해결

### ETC
- FCM 이용 시 발생할 수 있는 예외를 어떻게 처리하는 것이 이상적일 지 의견을 듣고 싶습니다.
- 채팅 관련 알림 로직은 추후 추가 예정

## 레퍼런스
<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->
- [Github Actions에서 비밀 사용법](https://docs.github.com/ko/actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow)
- [알림 서비스로 시작하는 서버 개발](https://youtu.be/CmTO68I2HSc?si=-6ZXl_HfQB6W8ngF)

Minjae-An committed:
feat: 알림 종류 enum 정의(#66)

알림 종류는 다음과 같다.
- 화살 수신
- 채팅 요청 수신
- 요청한 채팅 수락됨

Minjae-An committed:
feat: 알림 엔티티 구성(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 조회 로직 정의(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 조회 로직 구현(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 조회 비즈니스 로직 구성(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 응답 DTO 정의(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 조회 API 구성(#66)

Minjae-An committed:
feat: 읽지 않은 알림 개수 조회 메서드 파라미터명 수정(#66)

Minjae-An committed:
feat: 알림 목록 조회 로직 정의(#66)

Minjae-An committed:
feat: 알림 목록 조회 로직 구현(#66)

Minjae-An committed:
feat: 알림 목록 응답 DTO 정의(#66)

Minjae-An committed:
feat: 알림 목록 조회 요청 DTO 정의(#66)

Minjae-An committed:
feat: 알림 목록 페이지 응답 DTO 정의(#66)

Minjae-An committed:
feat: 알림 목록 조회 로직 추가(#66)

Minjae-An committed:
feat: 받은 알림 목록 조회 비즈니스 로직 추가(#66)

페이지 형태로 알림 목록 제공, 페이지 사이즈는 고정 9

Minjae-An committed:
feat: 받은 알림 목록 조회 API 추가(#66)

query string 통해 페이지 번호를 받아 받은 알림 목록 제공

Minjae-An committed:
feat: 알림 동의 내역 찾지 못함 예외 정의(#66)

Minjae-An committed:
feat: 알림 동의 내역 엔티티 구성(#66)

- 알림 타입의 경우 외부에서 요청하여 생성하는 것이 아닌 코드에서만 설정
- 열거형 타입으로 정의하여 코드에서 안정적으로 다룰 수 있도록 구성
- permissionStatus의 경우 @Getter 사용시 is... 네이밍으로 형성되어
  의미를 명확히 표현하지 못한다고 생각해 별도의 getter 구성

Minjae-An committed:
feat: 사용자, 알림 종류 기반 알림 동의 내역 조회 로직 정의(#66)

Minjae-An committed:
feat: 사용자 알림 동의 내역 조회 로직 구성(#66)

사용자의 알림 동의 내역중 주어진 알림 종류에 해당하는 내역 조회

Minjae-An committed:
feat: 사용자 알림 동의 내역 조회 비즈니스 로직 구성(#66)

Minjae-An committed:
feat: 사용자 알림 동의 내역 조회 응답 DTO 구성(#66)

Minjae-An committed:
feat: 알림 동의 내역 조회 API 구성(#66)

Minjae-An committed:
feat: 사용자 알림 동의 내역 수정 요청 DTO 수정(#66)

- 요청한 채팅 수락 시 알림 동의 여부 필드 추가
- not null 제약 조건 추가

Minjae-An committed:
feat: 사용자 알림 동의 내역 수정 비즈니스 로직 구성(#66)

더티 체킹 활용, 사용자의 각 종류별 알림 동의 내역의 동의 상태 업데이트

Minjae-An committed:
feat: 알림 동의 내역, 동의 상태 수정 메서드 추가(#66)

Minjae-An committed:
feat: 사용자 알림 동의 내역 수정 API 구성(#66)

Minjae-An committed:
feat: fcm 사용 설정 추가(#66)

Minjae-An committed:
feat: fcm 이용, 알림 전송 로직 추가(#66)

알림 제목, 내용외 부가적인 데이터로 다음과 같은 것들이 포함된다.
- 알림 생성 사용자 ID(화살 보낸 사용자, 채팅 요청한 사용자 등)
- 알림 생성 일시

Minjae-An committed:
feat: 알림 전송 이벤트 DTO 정의(#66)

- 모든 필드가 불변이기에 레코드 활용
- 알림 제목, 내용을 제공하는 편의 메서드 구성

Minjae-An committed:
feat: 알림 내역 저장 로직 추가(#66)

Minjae-An committed:
feat: FCM 커스텀 예외 정의(#66)

Minjae-An committed:
feat: 사용자 엔티티, device token 필드 추가(#66)

- 알림 전송을 위한 device token 필드 추가
- device token 갱신 로직 추가

Minjae-An committed:
feat: 알림 타입 열거형에 알림 메시지 필드 추가(#66)

알림 타입마다 메시지 형식이 정해져 있기에 타입별로 메시지를
함께 관리할 수 있도록 메시지 필드 추가

Minjae-An committed:
feat: 알림 내역 생성시, 읽음 여부 기본 false 설정(#66)

Minjae-An committed:
feat: 알림 내역 저장 로직 추가(#66)

Minjae-An committed:
feat: 알림 전송 이벤트 핸들링 로직 추가(#66)

알림 전송 이벤트 발생시 다음 두 작업을 수행한다.
1. FCM을 이용해 사용자 기기에 푸시 알림 전송
2. 알림 내역 저장

Minjae-An committed:
chore: fcm 관련 리소스 ignore 설정(#66)

Minjae-An committed:
chore: fcm admin sdk 의존성 추가(#66)

Minjae-An committed:
feat: 잘못된 필드명, 변수명 수정(#71)

sentUser -> sender, receivedUser -> receiver로 수정

Minjae-An committed:
feat: 화살 내역 존재 확인 메서드명 수정(#71)

엔티티 필드명 수정에 따른 변경

Minjae-An committed:
feat: 화살 내역 존재 확인 메서드 수정(#71)

엔티티 필드명 수정에 다른 파라미터명, 메서드명 변경

Minjae-An committed:
feat: 엔티티 필드명 수정에 따른 변경(#71)

Minjae-An committed:
feat: 화살 보내기 로직, 파라미터명 수정(#71)

Minjae-An committed:
Merge branch 'fix/#71-modify-arrow-transaction-field-name' into feat/#66-notification

Minjae-An committed:
feat: 읽지 않은 알림 존재 여부 응답 DTO 추가(#66)

Minjae-An committed:
feat: 읽지 않은 알림 존재 여부 조회 로직 추가(#66)

Minjae-An committed:
feat: 읽지 않은 알림 존재 여부 조회 로직 추가(#66)

Minjae-An committed:
feat: 읽지 않은 알림 존재 확인 로직 추가(#66)

Minjae-An committed:
feat: 읽지 않은 알림 존재 확인 API 추가(#66)

Minjae-An committed:
feat: 푸시 알림 전송 요청시, 알림에 표시될 이미지 포함(#66)

Minjae-An committed:
feat: 화살 보내기 로직, 화살 수신 이벤트 발행 코드 추가(#66)

Minjae-An committed:
Merge branch 'dev' into feat/#66-notification

Minjae-An committed:
feat: 알림 읽음 처리 로직 추가(#66)

Minjae-An committed:
feat: 사용자가 받은 알림 목록 조회 로직 추가(#66)

Minjae-An committed:
feat: 사용자가 받은 알림 목록 조회 로직 추가(#66)

- 기존 DTO 형태 조회 로직 삭제
- 엔티티로 받은 알림 목록을 조회하는 로직 추가

Minjae-An committed:
feat: 사용하지 않는 코드 삭제(#66)

Minjae-An committed:
feat: 알림 목록 조회 응답 DTO 수정(#66)

- 알림 ID 필드 추가
- 편의성 팩토리 메서드 추가

Minjae-An committed:
feat: 알림 목록 조회 페이지 응답 DTO, 팩토리 메서드 추가(#66)

Minjae-An committed:
feat: 받은 알림 목록 조회시 조회되는 알림 읽음 처리 추가(#66)

- 받은 알림 목록의 조회할 경우 읽지 않은 알림은 전부 읽음 처리되어야 함
- 이에 따라 비즈니스 로직에 로직 추가

Minjae-An committed:
feat: 알림 관련 API, URL에 /users 접두사 추가(#66)

사용자에 종속된 알림 목록에 접근한다는 점을 잘 표현할 수 있도록 URL 일괄 수정

Minjae-An committed:
feat: device token 업데이트 요청 DTO 추가(#66)

Minjae-An committed:
feat: device token 업데이트 로직 추가(#66)

Minjae-An committed:
feat: device token 업데이트 API 추가(#66)

Minjae-An committed:
feat: 불필요한 FCM 예외 삭제(#66)

Minjae-An committed:
feat: 알림 전송 이벤트 구성 데이터 변경(#66)

Minjae-An committed:
feat: device token not found 예외 추가(#66)

Minjae-An committed:
feat: device token not found 예외 처리 로직 추가(#66)

Minjae-An committed:
feat: 알림 전송 이벤트 변경에 따른 사용 위치 수정(#66)

Minjae-An committed:
feat: 화살 보내기시 알림 발행 로직 수정(#66)

Minjae-An committed:
feat: 비동기 사용 설정 정의(#66)

Minjae-An committed:
feat: 채팅 관련 상수 수정, 알림 메시지 제공 메서드 추가(#66)

Minjae-An committed:
feat: 사용하지 않는 응답 DTO 삭제(#66)

Minjae-An committed:
feat: 필드 타입, 기본 타입으로 변경(#66)

Minjae-An committed:
feat: 알림 메시지 제공 메서드 수정(#66)

알림 메시지 제공 주체 변경에 따라 수정

Minjae-An committed:
feat: badge 조회 로직 추가(#66)

- ios에서 앱에 표기되는 뱃지는 사용자가 확인하지 않은 알림, 메시지 등을 의미
- 기존 읽지 않은 알림 개수를 조회하는 로직은 이를 위해 필요
- 하지만, 뱃지는 별도의 API가 아닌 푸시 알림 데이터에 포함되어야 하는 것으로 파악
- 이에 불필요한 비즈니스 로직을 삭제하고, 단순히 뱃지를 조회하여 제공하는 로직 추가

Minjae-An committed:
feat: 알림 타입 상수 이름 변경에 따른 사용 위치 수정(#66)

Minjae-An committed:
feat: badge 조회 로직 추가(#66)

- 알림 전송 이벤트는 생성자, 수신자, 알림 타입, 알림 생성 일시를 저장
- badge의 경우 fcm을 통해 푸시 알림을 보낼 때만 필요하므로 이벤트 데이터와 분리
- 이벤트 리스너에서 badge로 조회하고 fcm을 통해 푸시 알림 전송시 전달하도록 설계

Minjae-An committed:
feat: 알림 전송 로직 변경(#66)

- Apns를 이용할 때 보다 구체화된 설정을 제공하는 ApnsConfig를 이용하도록 로직 수정
- 알림에 badge 포함, badge의 의미는 알림 수신자가 읽지 않은 알림 개수
- 알림에 이미지 포함하는 로직 삭제
- fcm 이용 도중 발생한 장애가 비즈니스 로직에까지 전파되는 형태는 부적절하다고 판단
- 다만, 외부 서비스와 관련하여 발생한 장애를 클라이언트에게 전달할 필요는 없다고 생각
- 서버 내부에서 관련 예외를 로그로 남겨, 장애 발생시 대응할 수 있는 형태로 예외 처리 구성

Minjae-An committed:
feat: 불필요한 API 삭제(#66)

Minjae-An committed:
feat: 알림 타입 일치 확인 메서드 추가(#66)

Minjae-An committed:
feat: 알림 동의 내역 목록 조회 메서드 추가(#66)

- 불필요한 조회 메서드 삭제
- 알림 동의 내역 목록 조회 메서드 추가

Minjae-An committed:
feat: 사용자 알림 동의 내역 목록 조회 로직 추가(#66)

Minjae-An committed:
feat: 알림 동의 내역 연관 로직 수정(#66)

- 알림 동의 내역 조회 및 수정시 목록 단위 작업토록 변경
- 알림 동의 내역 수정시 특정 알림 타입의 내역이 존재하지 않으면 새로 생성
- 이외에는 기존 동의 내역의 동의 상태를 수정하도록 로직 변경

Minjae-An committed:
feat: 알림 on/off 설정 API, http method PUT으로 변경(#66)

- 해당 API에 요청을 통해 새로운 알림 동의 내역을 생성할 수 있음
- 리소스가 존재하지 않으면 새로 생성한다는 측면에서 PUT으로 http method 변경

Minjae-An committed:
feat: 불필요한 읽음 여부 필드 삭제(#66)

알림 목록 조회시 조회되는 알림들을 전부 읽음 처리하기에 읽음 여부 필드 삭제

Minjae-An committed:
feat: 알림 목록(페이지) 요청 DTO 수정(#66)

- DTO 이름, 알림 목록(페이지)을 요청한다는 의미를 잘 나타내도록 수정
- 요청 바디를 통해 DTO 필드를 매핑함에 따라 생성자 및 필드 명세 수정
- 필드 타입, 기본 타입으로 수정

Minjae-An committed:
feat: 알림 목록 조회 로직 변경(#66)

- 알림 목록을 조회할 시 가장 최근에 받은 알림부터 조회되어야 함
- 생성 일시 내림차순 기준으로 알림 목록을 조회하도록 로직 변경

Minjae-An committed:
feat: 알림 목록 조회 로직 변경(#66)

- 최근에 받은 알림 목록을 조회한다는 의미를 잘 표현하게 메서드명 수정
- 조회 로직 변경에 따라 코드 수정

Minjae-An committed:
feat: 알림 목록 조회 메서드명, 로직 수정(#66)

- 최근에 받은 알림 목록 조회한다는 의미를 표현하도록 메서드명 수정
- 조회 메서드명 변경에 따른 사용 위치 수정

Minjae-An committed:
feat: 알림 목록 조회 API 수정(#66)

- 가장 최근에 받은 알림부터 조회된다는 의미를 표현할 수 있게 메서드명 수정
- 알림 목록 조회시 조회되는 알림들이 읽음 처리됨
- 해당 API가 멱등성을 보장하지 않으므로 PATCH로 http method 변경
- PATCH 사용시 요청 데이터는 request body를 통해 전달되는 방식이 권장됨
- 이에 따라 요청 DTO를 request body를 통해 생성하도록 로직 수정

Minjae-An committed:
feat: 알림 동의 내역 조회 로직 추가(#66)

Minjae-An committed:
feat: 알림 동의 내역 조회 로직 추가(#66)

Minjae-An committed:
feat: 사용자 알림 동의 여부 확인 로직 추가(#66)

Minjae-An committed:
feat: 사용자 알림 동의 여부 확인 로직 추가(#66)

특정 타입의 알림을 사용자가 동의했을 경우만 푸시 알림으로 전달하도록 로직 추가

Minjae-An committed:
feat: 최근 받은 알림 목록 조회 메서드명 수정(#66)

Minjae-An committed:
feat: 알림 동의 내역 업데이트 로직 수정(#66)

알림 타입, 알림 동의 상태로 구성된 Map을 활용하여 중복 로직 제거

Minjae-An committed:
feat: device token null 여부 검증 코드 제거(#66)

Minjae-An committed:
feat: 알림 전송 가능 확인 로직에 device token 검증 로직 추가(#66)

- 기존 코드에서는 fcm 유틸 클래스의 푸시 알림을 보내는 로직에서 device token 검증
- 해당 로직은 비동기로 처리되기 때문에 예외 처리가 여러움
- 클래스 책임면에서 해당 검증은 유틸 클래스에서 수행하는 것이 부적절하다고 판단
- 특정 사용자에게 알림 전송이 가능한 지 확인하는 로직에서 device token 존재 여부도 판단

Minjae-An committed:
feat: 알림 전송 가능 여부 확인 메서드명 변경(#66)

Minjae-An committed:
feat: 알림 동의 내역 저장 로직 추가(#66)

Minjae-An committed:
chore: fcm account key 설정 스텝 추가(#66)

Minjae-An committed:
chore: resources/fcm 디렉토리 생성 커맨드 추가(#66)

Minjae-An committed:
chore: fcm account key base64 decode 과정 추가(#66)

- github actions는 json 파일을 직접 secret으로 관리하는 형태를 권장하지 않음
- json 파일을 base64 인코딩된 형태로 secret으로 관리
- actions 수행시 디코딩하여 사용하는 형식으로 수정

Minjae-An committed:
chore: fcm account key 환경 변수 설정 구절 추가(#66)

Minjae-An committed:
chore: FCM_ACCOUNT_KEY 환경 변수 디코딩 구절 수정(#66)

Minjae-An committed:
feat: 알림 내역 저장 후 푸시 알림 전송토록 수정(#66)

Minjae-An committed:
Merge branch 'dev' into feat/#66-notification

Minjae-An committed:
feat: 응답에 포함된 알림 생성자 이름, 닉네임으로 수정(#66)

Minjae-An committed:
feat: 알림 메시지에 알림 생성자 이름이 아닌 닉네임 포함토록 수정(#66)

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/entity/NotificationType.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+@AllArgsConstructor
+public enum NotificationType {
+
```
해당 클래스를 엔티티 패키지 하위에 위치시킨 이유가 무엇인가요?

Minjae-An replied:
해당 열거형은 알림 엔티티를 구분하는 용도로 사용되며 해당 엔티티와 관련된 유일한 열거형이기 때문에 엔티티와 같은 패키지 내에 포함되어 있는 것이 관리가 편리하다고 판단했습니다. 하지만 `Gender`,`LoginType`과 같은 경우와 마찬가지로 별도의 `notification.enums` 패키지에 관리하는 것이 더 적합하다고 판단됩니다. 수정하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/event/NotificationSendEvent.java:
```diff
@@ -0,0 +1,22 @@
+import yeonba.be.notification.entity.NotificationType;
+import yeonba.be.user.entity.User;
+
+public record NotificationSendEvent(
+    NotificationType type,
```
record를 사용한 이유에 대해서 알려주세요.

Minjae-An replied:
- `record`를 사용할 경우 객체의 불변성을 보장하며 데이터 운반 목적의 클래스를 구성하는데 유용합니다.
- `getter`, `equals()`와 같은 코드를 자동으로 생성해주어 lombok을 활용한 기존 코드보다 훨씬 코드가 간결해집니다.
- 이벤트 객체의 경우 순전히 로직간 데이터 전달 용도로 사용되므로 `record`를 활용하면 불변성을 보장하며 그 용도를 더 확실히 표현할 수 있다고 판단하였습니다.

acceptor-gyu commented on .gitignore:
```diff
@@ -204,3 +204,4 @@ gradle-app.setting
 
 .idea
 be/src/main/resources/application.properties
+be/src/main/resources/fcm/*
```
```suggestion
be/src/main/resources/fcm/*

```

Minjae-An replied:
작성에 혼동이 있었습니다. 바로 수정하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/service/MyPageService.java:
```diff
@@ -147,4 +166,85 @@ private void uploadProfilePhotos(List<MultipartFile> profilePhotos, User user) {
+        // 알림 타입, 알림 동의 내역 Map 구성
+        Map<NotificationType, NotificationPermission> typePermissionMap =
+            notificationPermissions.stream()
+                .collect(Collectors.toMap(NotificationPermission::getType, Function.identity()));
+
+        // 알림 타입, 알림 동의 상태 Map 구성
+        Map<NotificationType, Boolean> typePermissonStatusMap = new HashMap<>();
+        typePermissonStatusMap.put(ARROW_RECEIVED, request.isAllowArrowReceivedNotification());
+        typePermissonStatusMap.put(CHATTING_REQUESTED,
+            request.isAllowChattingRequestNotification());
+        typePermissonStatusMap.put(
+            CHATTING_REQUEST_ACCEPTED, request.isAllowChattingRequestAcceptedNotification());
+
+        // 알림 타입별 내역 수정 or 내역 생성 작업 수행
+        typePermissonStatusMap.forEach((type, permissionStatus) ->
+            updateOrCreateNotificationPermissionBy(typePermissionMap, user, type, permissionStatus)
+        );
+    }
+
+    private void updateOrCreateNotificationPermissionBy(
+        Map<NotificationType, NotificationPermission> typePermissonMap,
+        User user,
+        NotificationType type,
+        boolean permissionStatus) {
+
+        Optional<NotificationPermission> foundNotificationPermission =
+            Optional.ofNullable(typePermissonMap.get(type));
+
+        // 동의 내역이 존재할 경우 동의 상태 업데이트
+        if (foundNotificationPermission.isPresent()) {
+            foundNotificationPermission.get().updatePermissionStatus(permissionStatus);
+
+            return;
+        }
+
+        // 동의 내역이 존재하지 않을 시, 생성 후 저장
+        NotificationPermission notificationPermission =
+            new NotificationPermission(permissionStatus, type, user);
+        notificationPermissionCommand.save(notificationPermission);
+    }
```
NotificationPermission을 User마다 회원 가입 시점에 status true로 다 등록을 하고 해당 로직에서는 List<NotificationPermission> list에서 request의 type에 매칭되는 status로 업데이트만 하는 것은 어떤가요?

Minjae-An replied:
프론트와 논의 끝에 다음과 같이 수정하기로 결정하였습니다. 
- 회원 가입 과정에서 일괄적으로 `true`처리
- 해당 로직에서는 업데이트만 수행

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/dto/request/NotificationPageRequest.java:
```diff
@@ -0,0 +1,22 @@
+        type = "number",
+        description = "페이지 번호",
+        example = "0")
+    @NotNull(message = "페이지 번호는 반드시 포함되어야 합니다.")
+    @PositiveOrZero(message = "페이지 번호는 0 이상이어야 합니다.")
+    private int page;
```
왜 필수 값이어야 하나요?

Minjae-An replied:
- 알림 목록 조회시 대량 데이터 조회가 발생할 수 있기에 페이지 형태로 응답을 제공합니다.
- 사용자가 어느 시점의(몇 페이지) 알림 목록을 조회하길 원하는 지 알아야 하기에 필수 값으로 페이지 번호를 받습니다.

acceptor-gyu replied:
페이지 형태로 응답하는 것은 맞는데, page 파라미터가 없는 경우 default page를 설정할 수 있지않나요?

Minjae-An replied:
제가 생각이 짦았네요, optional하게 요청할 수 있도록 수정하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/repository/NotificationQuery.java:
```diff
@@ -0,0 +1,31 @@
+        return notificationRepository.existsByReceiverAndReadIsFalse(receiver);
+    }
+
+    public Page<Notification> findRecentlyReceivedNotificationsBy(
+        User receiver, PageRequest pageRequest) {
+
+        return notificationRepository.findAllByReceiverOrderByCreatedAtDesc(receiver, pageRequest);
+    }
```
이후에 읽음 처리를 하는 로직이 있는 것같은데, 이렇게 하면 읽지 않은 알림이 많은 경우 사용자는 알림 조회를 했지만 최근 알림만 읽음 처리가 되고 오래된 알림은 남아있게 될 것같은데 제가 이해한 게 맞을까요?

Minjae-An replied:
로직이 잘못 되었다고 판단됩니다. 일단 저는 사용자 조회 일시 이전에 수신한 알림을 알림 목록 조회시 전부 읽음 처리하는 방향으로 로직을 개선하려 하는데 혹시 더 좋은 방향이 있다면 의견 주시면 감사하겠습니다.

acceptor-gyu replied:
읽음처리되지 않은 알림들에 대해서 전부 읽음 처리하도록 하면 될 것같습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/service/NotificationService.java:
```diff
@@ -0,0 +1,100 @@
+
+    private void readNotifications(List<Notification> notifications) {
+
+        notifications.stream()
+            .filter(notification -> !notification.isRead())
+            .forEach(Notification::read);
```
아래 코드와 결과적으로 차이가 없을 것같은데 filter()가 들어가 있는 이유는 무엇인가요?

```suggestion
        notifications.stream()
            .forEach(Notification::read);
```

Minjae-An replied:
불필요한 `UPDATE`쿼리가 발생하지 않도록 읽지 않은 알림을 필터링하는 로직을 추가하였습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/service/NotificationService.java:
```diff
@@ -0,0 +1,100 @@
+        long receiverId, NotificationPageRequest request) {
+
+        int pageNumber = request.getPage();
+        int size = 9;
+
```
size를 9로 고정시킨 이유는 무엇인가요?

Minjae-An replied:
요구사항에 따라 최근 알림 목록 조회시 고정 9개의 알림을 한 페이지에 담아 응답하도록 정하였습니다.

acceptor-gyu replied:
기획 명세서의 요구사항은 한 화면에 보이는 9개인 거고, UI상 페이지를 클릭하면서 넘어가는 방식이 아니라 스크롤링인데 서버에서도 9개씩만 넘겨줘야 한다고 생각하신 건가요?

Minjae-An replied:
확실히 데이터를 9개씩 넘기는 것은 너무 빈번하게 요청을 하게 만드는 구조인 듯 합니다. 프론트와 협의하여 한 번에 넘겨주는 데이터 수를 늘리겠습니다.

acceptor-gyu reviewed (commented):
모든 코드를 디테일하게 보지는 못했고 눈에 띄는 부분만 일단 리뷰했습니다. 확인 부탁드립니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/dto/response/NotificationResponse.java:
```diff
@@ -0,0 +1,74 @@
+        example = "민재님이 화살을 보냈어요!")
+    private String content;
+
+    @Schema(
+        type = "number",
+        description = "알림 보낸 사용자 ID",
+        example = "1")
+    private long creatorId;
+
+    @Schema(
+        type = "number",
+        description = "알림 보낸 사용자 대표 프로필 사진 URL",
+        example = "profile-photo/1-0")
+    private String creatorProfilePhotoUrl;
+
+    @Schema(
+        type = "string",
+        description = "알림 보낸 사용자 별명",
+        example = "안민재")
+    private String creatorNickname;
```
creator라는 것은 server 입장에서 요청을 보내 notification을 생성했다는 의미로 사용하신 것같은데, 이것을 client 입장에서는 이해하기 쉽지 않을 것같아서 creator라는 단어를 사용하는 것이 좋지 않아보입니다. sender와 같은 단어는 어떤가요?

Minjae-An replied:
너무 비즈니스 로직적인 맥락만 생각한 것 같습니다. 반영하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/controller/NotificationController.java:
```diff
@@ -0,0 +1,56 @@
+            .body(new CustomResponse<>(response));
+    }
+
+    @Operation(summary = "읽지 않은 알림 존재 여부 조회", description = "읽지 않은 알림 존재 여부 확인 가능")
+    @ApiResponse(responseCode = "200", description = "읽지 않은 알림 존재 여부 확인 성공")
+    @GetMapping("/users/notifications/unread/exists")
+    public ResponseEntity<CustomResponse<NotificationUnreadExistResponse>>
+    getUnreadNotificationExistence(@RequestAttribute("userId") long userId) {
+
+        NotificationUnreadExistResponse response =
+            notificationService.isUnreadNotificationExist(userId);
+
+        return ResponseEntity
+            .ok()
+            .body(new CustomResponse<>(response));
+    }
```
읽지 않은 알림의 존재 여부는 왜 필요한 것인가요?

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/dto/response/NotificationPageResponse.java:
```diff
@@ -0,0 +1,56 @@
+        example = "100")
+    private long totalElements;
+
+    @Schema(
+        type = "boolean",
+        description = "첫 페이지 여부",
+        example = "false")
+    private Boolean isFirst;
+
+    @Schema(
+        type = "boolean",
+        description = "마지막 페이지 여부",
+        example = "true")
+    private Boolean isLast;
```
Wrapper 타입인 이유는 무엇인가요?

Minjae-An replied:
JSON 필드 이름때문에 이렇게 구성하였는데 `@JsonProperty`를 활용하는 방향으로 변경하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/notification/dto/response/NotificationResponse.java:
```diff
@@ -0,0 +1,74 @@
+@AllArgsConstructor
+public class NotificationResponse {
+
+    @Schema(
+        type = "number",
+        description = "알림 ID",
+        example = "1")
+    private long id;
```
response에 notificationId가 필요한 이유는 무엇인가요?

Minjae-An replied:
처음 로직을 구성할 때 프론트에서 알림 목록을 정렬하는데 필요한 단순한 기준 데이터가 필요하다고 판단하여 제공했었습니다. 하지만, 이미 정렬된 형태로 데이터를 제공하고 있으므로 제거하겠습니다.

Minjae-An committed:
chore: EOF 추가(#66)

Minjae-An committed:
feat: 알림 타입 열거형 패키지 변경(#66)

- 알림 타입 열거형, notification.enums 패키지로 이동
- 사용 위치 수정

Minjae-An committed:
feat: 페이지 번호 nullable하도록 변경(#66)

- 페이지 번호를 nullable하도록 변경
- 페이지 번호를 요청에 포함하지 않을 경우 기본 첫 페이지 제공

Minjae-An committed:
feat: 알림 엔티티 필드명 수정, 읽음 처리 메서드 삭제(#66)

- creator 필드명, sender로 수정
- 사용하지 않는 읽음 처리 메서드 삭제

Minjae-An committed:
feat: ID 기반 알림 읽음 처리 로직 정의(#66)

Minjae-An committed:
feat: ID 기반 알림 읽음 처리 로직 구현(#66)

- 벌크성 수정 쿼리를 통해 특정 알림 ID보다 ID가 작거나 같은 알림들 전부 읽음 처리
- 벌크성 수정 쿼리는 영속성 컨텍스트를 무시, DB와 컨텍스트의 엔티티 상태가 달라질 수 있음
- 이를 방지하기 위해 수정 쿼리 실행 이후 영속성 컨텍스트 초기화

Minjae-An committed:
feat: creator 변수명, sender로 수정(#66)

알림을 '보내는 사용자'라는 의미를 잘 표현하기 위해 변수명 수정

Minjae-An committed:
feat: 첫 페이지, 마지막 페이지 여부 필드 기본 타입으로 변경(#66)

Minjae-An committed:
feat: 커스텀 repository 인터페이스 상속 추가(#66)

Minjae-An committed:
feat: ID 기반 알림 읽음 처리 로직 추가(#66)

Minjae-An committed:
feat: 최근 받은 알림 목록 조회 로직 수정(#66)

- 한 번에 제공되는 알림 데이터 수, 45개로 조정
- 기존 로직의 경우 첫 페이지에서 조회되지 않은 알림이 계속 읽음 처리되지 않을 수 있는 오류 존재
- 알림 목록 조회시 가장 최근에 받은 알림 ID 기반, 그 이전 알림 전부 읽음 처리하도록 로직 수정

Minjae-An committed:
Merge branch 'dev' into feat/#66-notification

Minjae-An committed:
feat: 알림 타입, 동의 여부 맵 제공 로직 추가(#66)

요청 DTO에 포함된 알림 타입별 동의 여부를 맵으로 제공할 수 있는 로직 추가

Minjae-An committed:
feat: 알림 동의 내역 생성시 무조건 동의 처리(#66)

Minjae-An committed:
feat: 알림 동의 내역, 리스트 단위로 저장하도록 로직 변경(#66)

Minjae-An committed:
feat: 동의 처리된 알림 동의 내역 생성 로직 추가(#66)

Minjae-An committed:
feat: 회원 가입시 모든 알림 허용 처리(#66)

- 회원 가입시 모든 알림 허용 처리
- 추후에 사용자가 수정할 수 있게 제공

Minjae-An committed:
feat: 알림 동의 내역 업데이트 로직 수정(#66)

- 요청 DTO를 통해 알림 타입별 동의 여부를 Map 형태로 치환
- 이 Map을 기반으로 조회된 알림 내역들, 타입별로 동의 여부 업데이트

Minjae-An committed:
feat: 알림 on/off 설정 API, PATCH로 수정(#66)

- 기존에는 해당 API를 통해 알림 내역이 없을 경우 생성
- 알림 내역 동의 여부를 수정만 하는 방향으로 로직 변경, 이에 따라 PUT에서 PATCH로 수정

Minjae-An committed:
feat: 알림 동의 내역 조회 로직 수정(#66)

알림 동의 내역 리스트를 바탕으로 알림 타입별 동의 내역 Map을 형성하여 응답하도록
코드를 간결하게 수정

Minjae-An committed:
feat: 불필요한 알림 ID 삭제(#66)

Minjae-An committed:
chore: EOF 추가(#66)

Minjae-An committed:
feat: 알림 읽음 처리 메서드명 변경(#66)

Minjae-An committed:
feat: 알림 읽음 처리 메서드 수정(#66)

- 기존 로직은 주어진 ID보다 작거나 같은 ID를 가진 알림을 전부 읽음 처리
- 모든 사용자가 받은 알림이 읽음 처리되는 잘못된 로직
- 특정 사용자가 받은 알림만 ID를 기준으로 알림 처리할 수 있도록 로직 수정
- 메서드명 변경

Minjae-An committed:
feat: 알림 읽음 처리 로직 수정(#66)

Minjae-An committed:
feat: 최근 알림 목록 조회 로직 수정(#66)

- 페이지 번호가 optional해짐에 따라 요청 DTO 검증 로직 추가
- 읽음 처리 메서드명 변경에 따른 사용 위치 수정

Minjae-An committed:
feat: 최근 받은 알림 목록 조회시 request body optional하게 수정(#66)

Minjae-An committed:
feat: 불필요한 줄바꿈 제거(#66)

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/controller/MyPageController.java:
```diff
@@ -96,11 +97,28 @@ public ResponseEntity<CustomResponse<Void>> updateProfile(
+    @Operation(summary = "알림 동의 내역(on/off) 조회", description = "알림 동의 내역을 조회할 수 있습니다.")
+    @ApiResponse(responseCode = "200", description = "알림 동의 내역 조회 성공")
+    @GetMapping("/users/notifications/permissions")
+    public ResponseEntity<CustomResponse<NotificationPermissionsResponse>>
+    getNotificationPermissions(@RequestAttribute("userId") long userId) {
```
```suggestion
    public ResponseEntity<CustomResponse<NotificationPermissionsResponse>>getNotificationPermissions(
        @RequestAttribute("userId") long userId) {
```

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/controller/MyPageController.java:
```diff
@@ -96,11 +97,28 @@ public ResponseEntity<CustomResponse<Void>> updateProfile(
+    public ResponseEntity<CustomResponse<NotificationPermissionsResponse>>
+    getNotificationPermissions(@RequestAttribute("userId") long userId) {
+
+        NotificationPermissionsResponse response =
+            myPageService.getNotificationPermissions(userId);
```
```suggestion
        NotificationPermissionsResponse response = myPageService.getNotificationPermissions(userId);
```

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/dto/request/UserAllowNotificationsRequest.java:
```diff
@@ -1,25 +1,46 @@
+        Map<NotificationType, Boolean> map = new HashMap<>();
+        map.put(NotificationType.ARROW_RECEIVED, this.allowArrowReceivedNotification);
+        map.put(NotificationType.CHATTING_REQUESTED, this.allowChattingRequestNotification);
+        map.put(NotificationType.CHATTING_REQUEST_ACCEPTED,
+            this.allowChattingRequestAcceptedNotification);
```
```suggestion
        map.put(NotificationType.CHATTING_REQUEST_ACCEPTED, this.allowChattingRequestAcceptedNotification);
```

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/dto/request/UserAllowNotificationsRequest.java:
```diff
@@ -1,25 +1,46 @@
+    @NotNull(message = "요청한 채팅 수락 시 알림 동의 여부는 필수 값입니다.")
+    private boolean allowChattingRequestAcceptedNotification;
+
+    public Map<NotificationType, Boolean> toNotificationTypePermissionStatusMap() {
+
+        Map<NotificationType, Boolean> map = new HashMap<>();
```
- Map을 사용하신 이유는 무엇인가요?
- 변수명에 특정 자료구조를 사용하는 것은 지양해주세요.

Minjae-An replied:
- 각 알림 타입에 해당하는 알림 동의 여부를 `Map`으로 다루는 것이 로직을 간결하게 구성할 수 있다고 
  판단하였습니다.
- 메서드명, 변수명 수정하겠습니다. `Map`타입의 변수 네이밍을 어떻게 해야할 지 감이 안 잡히는데 팁을 좀 주실 수 
  있을까요?

acceptor-gyu replied:
솔직히 이 기능을 위해서 사용된 로직이 간결해 보이지가 않습니다.(처음 만든 Map의 key, value가 각각 무엇을 의미하는지, DB에 저장할 엔티티는 언제 만들어지는 등 코드 읽기가 쉽지 않습니다.) 
Map을 로직에 사용하게 되면 간결하지 않은 경우가 많아지는 것같습니다.(제 경험상으로는 그렇습니다.) 
그래서 저는 해당 API request를 바꾸는 것이 더 좋아보입니다.
예를 들면 [{NotificationType, Status}] 형태로 받으면 단순 List로 훨씬 간결하게 코드를 작성할 수 있을 것같은데, 이렇게 하기에는 무리가 있을까요?

- 저는 Map이든 List든 크게 신경안쓰고 그 변수에 담겨지는 데이터 자체가 의미하는 게 무엇인지를 생각하고 최대한 그대로 표현하는 것이 아직까지 제 기준에서는 나쁘지는 않은 것같습니다.

Minjae-An replied:
`알림 유형`,`동의 여부` 필드를 가지는 DTO를 별도로 정의하여 로직을 재구성하였습니다. DTO 이름, 요청 값과 열거형 매핑, 알림 동의 내역 업데이트와 관련하여 더 좋은 방향이 있으면 의견 주시면 감사하겠습니다.
참고로, 요청 DTO의 `List` 타입 필드에 `null`값이 포함되어 있는 지 확인하는 검증은
```
List<@NotNull NotificationPermissionDetail> permissions;
```
위와 같은 형태로도 수행이 가능하지만, 예외 메시지를 추가하면 다소 가독성이 떨어진다고 생각하여 서비스 계층에서 
검증하고 커스텀 예외를 발생시키도록 로직을 구성했습니다.

Minjae-An committed:
feat: 변수명, 메서드명에서 Map 제외(#66)

Minjae-An committed:
feat: 변수명에서 Map 제외(#66)

Minjae-An committed:
refactor: 코드 정렬 수정(#66)

Minjae-An committed:
feat: 알림 동의 내역 상세 DTO 구성(#66)

Minjae-An committed:
feat: 알림 내역 업데이트 요청 DTO 변경(#66)

- 알림 내역 상세 DTO를 이용도록 변경
- 알림 내역에 초점을 두도록 클래스 이름 변경

Minjae-An committed:
feat: 알림 내역 조회 응답 DTO 변경(#66)

- 알림 내역 상세 DTO를 이용하도록 변경
- 알림에 초점을 두도록 클래스 이름 변경

Minjae-An committed:
feat: 알림 동의 내역 조회, 알림 동의 내역 업데이트 로직 수정(#66)

새로 구성한 알림 동의 내역 상세 DTO를 이용하여 코드를 보다 간결하게 변경

Minjae-An committed:
feat: 요청 동의 여부에 null 값이 포함되는 예외 정의(#66)

Minjae-An committed:
feat: 필드 검증 추가(#66)

Minjae-An committed:
feat: 필드 검증 추가(#66)

Minjae-An committed:
feat: 알림 동의 내역 업데이트 로직 수정(#66)

- 요청 동의 내역 리스트, null 값 포함 여부 검증 추가
- 요청 DTO, 알림 유형 필드 타입 변경에 따른 업데이트 로직 수정

Minjae-An committed:
feat: 알림 on/off 설정 API, 요청 DTO 이름 변경(#66)

Minjae-An committed:
feat: 요청 동의 내역에 null이 포함되는 예외, 메시지 수정(#66)

Minjae-An committed:
feat: 컬렉션 요소들에 대한 검증 추가(#66)

Minjae-An committed:
feat: 알림 유형 필드, 검증 수정(#66)

- 기존 정규식 검증으로도 공백 값 검증 가능
- 공백 값 요청될 시 좀 더 상세한 예외 상황 표현을 위해 @NotBlank 어노테이션으로 변경

acceptor-gyu reviewed (approved):
시연이 급해 approve 하겠습니다. 이후 버그 발생 시 수정부탁드립니다.

Minjae-An committed:
Merge branch 'dev' into feat/#66-notification
